kimchi/src/circuits/polynomials/range_check/gate.rs

```diff
@@ -126,19 +157,36 @@ impl<F: FftField + SquareRootField> CircuitGate<F> {
         let gamma = F::rand(rng);
         let z_poly = cs
             .perm_aggreg(&witness, &beta, &gamma, rng)
-            .map_err(|_| format!("Invalid {:?} constraint - permutation failed", self.typ))?;
+            .map_err(|_| GateError::InvalidCopyConstraint(self.typ))?;
 
         // Compute witness polynomial evaluations
         let witness_evals = cs.evaluate(&witness_poly, &z_poly);
 
+        let mut index_evals = HashMap::new();
+        index_evals.insert(
+            self.typ,
+            &cs.range_check_selector_polys[circuit_gate_selector_index(self.typ)].eval8,
+        );
+
+        // Set up lookup environment
+        let lcs = cs
+            .lookup_constraint_system
+            .as_ref()
+            .ok_or(GateError::MissingLookupConstraintSystem(self.typ))?;
+
+        let lookup_env_data =
+            set_up_lookup_env_data(self.typ, cs, &witness, &beta, &gamma).map_err(|e| e)?;
+        let lookup_env = Some(LookupEnvironment {
```

Comment:
Do you think it may be worth creating a builder for `LookupEnvironment`? Not for this PR, but for another one as part of the kimchi api epic.

Reply:
Yes, possibly. Great idea.  This is related to the suggestion by Matthew to move `set_up_lookup_env_data` to abstract out the `LookupContext` logic in prover.